### PR TITLE
Counters - First pass at implementation

### PR DESCRIPTION
**Pull request by bmeike**
Rewrite of the implementation of counters as MutableRealmIntegers.
Preliminary implementation of #4266

### CODE REVIEW

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 1]
```diff
@@ -556,6 +571,8 @@ private void emitCodeForUnderConstruction(JavaWriter writer, boolean isPrimaryKe
                 .emitEmptyLine();
     }
 
+    // Note that because of bytecode hackery, this method may run before the constructor!
+    // It may ever run before fields have been initialized.
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: ever - even

Comment (cmelchior) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: Also, I don't believe it can run _before_ the constructor,  but it will be invoked during the constructor. But I guess the effect is the same. Fields might not initialized.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyInterfaceGenerator.java [hunk 2]
```diff
@@ -54,24 +54,29 @@ public void generate() throws IOException {
...
+            String fieldName = field.getSimpleName().toString();
+            String fieldTypeCanonicalName = field.asType().toString();
+            writer
+                    .beginMethod(
+                            fieldTypeCanonicalName,
+                            metaData.getInternalGetter(fieldName),
+                            EnumSet.of(Modifier.PUBLIC))
+                    .endMethod();
+
+            // Counters do not have setters.
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyInterfaceGenerator.java]: Counters -> MutableRealmIntegers

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java [hunk 3]
```diff
@@ -94,7 +94,8 @@ public void compileProcessedEmptyFile() throws Exception {
                 .failsToCompile();
     }
 
-    // Disabled because it does not seem to find the generated interface file @Test
+    @Ignore("Disabled because it does not seem to find the generated interface file")
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java]: This looks a bit worrying?

File: realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 4]
```diff
@@ -662,6 +685,13 @@ public static String getTableName() {
                 ((AllTypesRealmProxyInterface) obj).realmSet$columnBinary(JsonUtils.stringToBytes(json.getString("columnBinary")));
             }
         }
+        if (json.has("columnCounter")) {
+            if (json.isNull("columnCounter")) {
+                throw new IllegalArgumentException("Trying to set non-nullable field 'columnCounter' to null.");
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Is this the semantics we want? By doing this, it looks impossible to initialize a counter to `null` from JSON?

Comment (cmelchior) [inline realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Also, in the streaming version we set the value to `null`, not throw?

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 5]
```diff
@@ -385,12 +392,18 @@ private boolean categorizeField(Element element) {
...
 
-        // Check @LinkingObjects last since it is not allowed to be either @Index, @Required or @PrimaryKey
+        // @LinkingObjects cannot be @PrimaryKey or @Index.
         if (field.getAnnotation(LinkingObjects.class) != null) {
+            // Do not add backlinks to fields list.
             return categorizeBacklinkField(field);
         }
 
-        // Standard field that appear valid (more fine grained checks might fail later).
+        // Similarly, a RealmInteger cannot be a @PrimaryKey, @Index or @LinkingObject.
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Also, any reason we cannot index it?

File: realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java [hunk 6]
```diff
@@ -59,228 +63,318 @@ public void tearDown() throws Exception {
...
+        assertEquals(r1, r2);
+
+        MutableRealmInteger r3 = r1;
+        r1.set(19);
+        assertEquals(19, r3.get().intValue());
+    }
+
+    @SuppressWarnings({"ReferenceEquality", "EqualsIncompatibleType"})
+    private void testEquality(Counters c1, Counters c2) {
+        c1.columnCounter.set(7);
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java]: Maybe `assertFalse(c1 == c2)` to clarify contract for using this method?

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+package io.realm;
+
+import io.realm.annotations.Beta;
+import io.realm.internal.ManagableObject;
+import io.realm.internal.Row;
+
+
+/**
+ * A MutableRealmInteger is a mutable, {@link java.lang.Long}-like numeric quantity.
+ * MutableRealmInteger implementations are not thread safe.  Like all Realm objects,
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: That only applies to unmanaged MutableRealmIntegers, no. Managed integers are thread-confined.

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+ * managed RealmIntegers may not be moved across threads.  Unmanaged RealmObject
+ * may be moved across threads but require safe publication.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: Not sure what "safe publication" means?

File: realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java [hunk 9]
```diff
@@ -59,228 +63,318 @@ public void tearDown() throws Exception {
...
+        assertTrue(n.equals(Long.valueOf(8)));
+        assertTrue(n.equals(c1.columnCounter.get()));
+        assertTrue(n.intValue() == c1.columnCounter.get().intValue());
+
+        c1.columnCounter.increment(1);
+        assertFalse(n.intValue() == c1.columnCounter.get().intValue());
+        assertFalse(n.intValue() == r1.get().intValue());
+    }
+
+    private void testNullability(Counters c1) {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java]: `testBasic`, `testEquality` and `testNullability` are all pretty complicated and seem to have some specific variants not really expressed in the name, but only halfway down the tests. Could assumptions about input be clarified in Javadoc for these methods?

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+    /**
+     * Managed Implementation.
+     */
+    // FIXME Counters: wire up to native increment method
+    static class Managed extends MutableRealmInteger {
+        private final ProxyState<?> proxyState;
+        private final BaseRealm realm;
+        private final Row row;
+        private final long columnIndex;
+        private Long value;
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: This looks dangerous. The value should be read from the `row` every time.

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+     * Decrements the {@code MutableRealmInteger}, subtracting the value of the argument.
+     * Increment/decrement from all devices are reflected in the new value, which is guaranteed to converge.
+     *
+     * @param dec quantity to be subtracted from the counter.
+     */
+    public abstract void decrement(long dec);
+
+
+    /**
+     * @return true if and only iff {@code get()} will return {@code null}.
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: iff -> if

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+        Long thisValue = get();
+        Long otherValue = o.get();
+        return (otherValue == null) ? ((thisValue != null) ? 1 : 0)
+                : (thisValue == null) ? ((otherValue != null) ? -1 : 0)
+                : thisValue.compareTo(otherValue);
+    }
+
+    /**
+     * A MutableRealmInteger's hash code depends only on its value.
+     * Will NPE if either value is null
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: Thats not true?

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+     */
+    @Override
+    public final int hashCode() {
+        Long thisValue = get();
+        return (thisValue == null) ? 0 : thisValue.hashCode();
+    }
+
+    /**
+     * Two RealmIntegers are {@code .equals} if and only if their longValues are equal.
+     * Will NPE if either value is null
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: Same?

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 12 above)
Comment (Zhuinden) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: I'd be surprised if `hashCode()` NPE'd o-o

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 13 above)
Comment (Zhuinden) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: I'd be surprised if `equals()` NPE'd o-O

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java [hunk 14]
```diff
@@ -479,6 +480,36 @@ public void compileWithInterfaceForObject() {
...
+        // FIXME Counters: Add test
+    }
+
+    @Test
+    public void failOnPKCounter() {
+        // FIXME Counters: Add test
+    }
+
+    @Test
+    public void failOnIndexedCounter() {
```
Comment (beeender) [inline realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java]: why? index should be allowed i think?

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 15]
```diff
@@ -385,12 +392,18 @@ private boolean categorizeField(Element element) {
...
+        if (Utils.isRealmInteger(field)) {
+            if (!categorizeMutableRealmIntegerField(field)) { return false; }
```
Comment (zaki50) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: To avoid returning when `categorizeMutableRealmIntegerField(field)` returns `true`?

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.java [hunk 16]
```diff
@@ -340,6 +353,40 @@ public void emitGetObjectWithPrimaryKeyValue(
...
+                   .beginControlFlow("if (json.isNull(\"%s\"))", fieldName)
+                       .emitStatement(Constants.STATEMENT_EXCEPTION_ILLEGAL_NULL_VALUE, fieldName)
+                   .nextControlFlow("else")
+                       .emitStatement("((%s) obj).%s().set(json.getLong(\"%s\"))", interfaceName, getter, fieldName)
+                   .endControlFlow()
+               .endControlFlow();
+       }
+       // @formatter:on
+
+        // @formatter:off
```
Comment (kneth) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.java]: Formatted incorrectly :smile: - a space too many

File: realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java [hunk 17]
```diff
@@ -59,228 +63,318 @@ public void tearDown() throws Exception {
...
+        ri.increment(17);
+        ri.decrement(7);
+        realm.commitTransaction();
+
+        assertEquals(Long.valueOf(42L), unmanagedRI.get());
+        assertEquals(Long.valueOf(47L), managedRI.get());
+    }
+
+    /**
+     * Assure that a MutableRealmInteger acquired from an unmanaged object is not affected by changes in the DB.
```
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java]: "DB" -> "Realm"

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 11 above)
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: or "if and only iff" -> "iff"

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+        return get() == null;
+    }
+
+
+    /**
+     * RealmIntegers compare strictly by their values.
+     * Null is a legal value for a MutableRealmInteger: null < non-null
+     *
+     * @param o the compare target
+     * @return -1, 0, or 1, depending on whether this object's value is &gt;, =, or &lt; the target's.
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: Won't it return `1` if greater-than?

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 11 above)
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: No, `iff` is not a good term if we want to make our documentation easy to read.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java (hunk 5 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: I'm not sure it is check completely, yet.  It will be though.  Perhaps a *FIXME* is in order...

As far as indexing, I don't actually see a reason to forbid it.  Seems odd to me but I can certainly allow it!

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java (hunk 15 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Yeah...  I think that would change the meaning.  What @zaki50 said.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java (hunk 1 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: That seems odd to me.  I have never seen a constructor see uninitialized fields.  I can guarantee that this method sometimes runs before fields are initialized.  I may be missing something but I don't see how it can run after the constructor, but before field initialization.  It is possible that it runs after superclass constructor.  Didn't check that.

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java (hunk 3 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java]: I would say so.  At least it isn't totally hidden, now.

File: realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java (hunk 4 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Yeah... I just roughed this stuff in.  There are no JSON tests yet.

File: realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java (hunk 6 above)
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java]: Yeah.  Good call.

File: realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java (hunk 9 above)
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java]: I can do this if you like, but I think it is a horrible idea.  As you well know, defining the semantics for this feature has been a long and painful discussion, often plagued by misunderstanding.  The idea that English which is never checked for accuracy, better represents the semantic definition, than the code, seems completely mistaken, to me.

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 7 above)
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: Exactly.  Neither one can be passed, safely, between threads.  Perhaps I should word that differently?

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 8 above)
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: Standard term.  "Java Concurrency in Practice", section 3.5.
tl;dr, just what it sounds like: you have to do some kind of synchronization to pass something to another thread.

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 10 above)
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: Yes, surely.  This is an intermediate commit.  I needed some way to make existing tests work.  Value gets removed.  Added FIXME comment

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 11 above)
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: I lost this discussion once before.  iff -> if.

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 12 above)
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: Ya.  As you can see, it won't NPE.  Copy pasta.  Deleted.

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 13 above)
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: Yeah.  Leftovers from the previous implementation.  Good catch.

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java (hunk 14 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java]: This is not in the spec.  Yes, I know that is because no spec exists.

Ok.  Will add to the spec.  Allowing both @Required and @Index seem like good choices.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.java (hunk 16 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.java]: Fixed

File: realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java (hunk 18 above)
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/MutableRealmInteger.java]: bleh!  Yes.  Good catch.

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java (hunk 3 above)
Comment (cmelchior) [inline realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java]: I would also add a FIXME, just so it isn't forgotten.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 19]
```diff
@@ -332,11 +337,13 @@ private boolean checkDefaultConstructor() {
 
     private boolean checkForFinalFields() {
         for (VariableElement field : fields) {
-            if (field.getModifiers().contains(Modifier.FINAL)) {
-                Utils.error(String.format(Locale.US,
-                        "Class \"%s\" contains illegal final field \"%s\".", className, field.getSimpleName().toString()));
-                return false;
-            }
+            if (!field.getModifiers().contains(Modifier.FINAL)) { continue; }
```
Comment (bmunkholm) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: please have `continue`and `return` on it's own line.

Comment (cmelchior) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: We introduced an automatic formatter that allows this kind of code. No-one on the team found it hindered readability, nor could lead to errors.

Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Sure.  I am the person that introduced the code formatter and I did run it on this code.

I believe that the team policy is that if the formatter accepts it, it is ok.  We can change the formatter, if you think it important.  The point, however, is that delaying a commit for yet another day, for an un-automated style complaint, is a significant impediment to cross-time-zone development.

**Comment by bmeike** [comment]
@cmelchior Sure.  IMO, that is the whole point of the @Ignore.  We should not have @Ignore d tests!  Every @Ignore documents a broken test, in a way that is visible *every* time you run the tests.

IMO, commenting out the @Test is criminal and should be punished with a severe lashing with a damp spaghetti noodle.  @Ignore -ing a test indicates that it is broken and needs to be fixed or removed.

That said, will add the FIXME

Re the merge, sorry.  I forgot to push the merged feature branch to origin.  Should be fixed now.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyInterfaceGenerator.java (hunk 2 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyInterfaceGenerator.java]: fixed

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java (hunk 3 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java]: ok!  done.

File: realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java (hunk 17 above)
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/RealmIntegerTests.java]: fixed.  you got your own personal commit for this...

**Comment by cmelchior** [comment]
Good point about `@Ignore` being very visible in the log. That is probably just my habit of searching for them 😄
